src/main/java/com/louis/HashTable.java

```diff
@@ -0,0 +1,272 @@
+package com.louis;
+
+import com.google.common.base.Preconditions;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+import java.lang.reflect.Array;
+import org.immutables.value.Value;
+import com.louis.HashEntry;
+import java.lang.String;
+import java.util.Optional;
+
+/**
+ * Implements a hashtable.
+ *
+ * Hashtables implement the associative array abstract data type. The tables contain keys and associated values.
+ * A hash function computes the location of a value for a given key.
+ *
+ * @author Louis Ashton (louisashton@live.com)
+ */
+public class HashTable {
+
+	private static final int MIN_CAPACITY = 32;
+	private static final double RESIZE_LOAD = 0.5;
+	private static final double MAX_LOAD = 0.75;
+	private static final double MIN_LOAD = 0.25;
+
+	private HashEntry[] table;
+	private int currentSize;
+	private int currentCapacity;
+	
+	/**
+	 * Creates a new HashTable.
+	 */
+	public HashTable() {
+		currentSize = 0;
+		currentCapacity = MIN_CAPACITY;
+		table = new HashEntry[MIN_CAPACITY];
+	}
+
+	/**
+	 * Returns true if key is a member of this table.
+	 *
+	 * @param key the key to check for in the table. May not be null.
+	 * @return Returns true if the table contains that key.
+	 * @throws IllegalArgumentException if key is null.
+	 */
+	public boolean contains(String key) {
+		Preconditions.checkArgument(key != null,
+				"First argument to contains(String key) is null");
+		int index = calculateDesiredPositionOfKey(key, currentCapacity);
+		return searchForKeyFromIndex(key, index).isPresent();
+	}
+
+	private int calculateDesiredPositionOfKey(String key, int modulus) {
+		return hash(key) % modulus;
+	}
+
+	private Optional<Integer> searchForKeyFromIndex(String key, int index) {
```

Comment:
Replace the body of this one with:

`return searchForKeyForwardOfIndex(key, index, (index - 1) % currentSize)`